src/search/parser/syntax_analyzer.cc

```diff
@@ -0,0 +1,249 @@
+#include "syntax_analyzer.h"
+
+#include "errors.h"
+#include "lexical_analyzer.h"
+
+#include <unordered_set>
+
+using namespace std;
+
+namespace parser {
+using std::to_string;
+
+static unordered_set<TokenType> literal_tokens {
+    TokenType::FLOAT,
+    TokenType::INTEGER,
+    TokenType::BOOLEAN,
+    TokenType::IDENTIFIER
+};
+
+class TokenContext {
+    const TokenStream &tokens;
+    const int token_lookahead;
+    std::vector <std::string> contexts;
+public:
+    explicit TokenContext(const TokenStream &tokens, int token_lookahead)
+        : tokens(tokens), token_lookahead(token_lookahead) {}
+
+    void push(const std::string &layer) {
+        contexts.push_back(layer + ": '" + tokens.dump(token_lookahead) + "'");
+    }
+
+    void pop() {
+        contexts.pop_back();
+    }
+
+    std::string str() const {
+        return utils::join(contexts, "\n");
+    }
+};
```

Comment:
I am not sure which state of the code you are reviewing. The TokenContext should be in the `syntax_analyzer.cc` file now and not accessible outside of it. It belongs to the error reporting. In the end it is something like a traceback in Python. If there is an error, it shows the steps that lead to the error, because without this information, it is hard to understand where the error happened and how to interpret it.

Reply:
Thanks! Perhaps Traceback would be a better name? Based on what is pushed into it, the part of the name that I understand least is *token* context. If it was called ParseContext or even Context, I probably would have had fewer question marks.